### PR TITLE
stmmac: add stmmac port

### PR DESCRIPTION
**Pull request by feraralashkar**
RFC

1. add stmmac netmap support (linux version), and

2. fix compiler warning when cross building the driver with:
arm-buildroot-linux-gnueabihf-gcc (Buildroot 2019.02.2-g0b3ebc6ca5-dirty) 8.3.0

### CODE REVIEW

File: LINUX/configure [hunk 1]
```diff
@@ -927,17 +927,21 @@ EOF
...
 				src=$(cd $src; pwd)
 				src_found=1
-			fi 
+			fi
 			src_checked=1
 		fi
 		drv_kernel_src=
 		if [ -n "$src_found" ]; then
-			drv_kernel_src=$(find $src/drivers/ -name "$d" -print -quit)
+			if [ "$d" = "stmmac_eth" ]; then
```
Comment (giuseppelettieri) [inline LINUX/configure:940]: We don't want to fill the script with special cases. Can't you just call the driver ```stmmac```?

File: LINUX/configure [hunk 2]
```diff
@@ -883,7 +883,7 @@ EOF
 		message " NOTE  " <<EOF
 We are trying to download the original sources for driver
 $d using the following command:
-  
+
```
Comment (giuseppelettieri) [inline LINUX/configure:886]: Please leave the whitespace alone, otherwise the patch touches too many unrelated lines.

File: LINUX/netmap_linux.c [hunk 3]
```diff
@@ -1105,6 +1105,12 @@ nm_os_generic_find_num_desc(struct ifnet *ifp, unsigned int *tx, unsigned int *r
 			*tx = netmap_generic_ringsize;
 		}
 		error = 0;
+	}else{
```
Comment (giuseppelettieri) [inline LINUX/netmap_linux.c:1108]: This is a generic function, we cannot put this stuff here. If you need these values please create your own ```nm_config``` callback.

File: sys/dev/netmap/netmap_mem2.c [hunk 4]
```diff
@@ -524,7 +524,7 @@ static struct netmap_obj_params netmap_min_priv_params[NETMAP_POOLS_NR] = {
 		.num  = 4,
 	},
 	[NETMAP_BUF_POOL] = {
-		.size = 2048,
+		.size = 2048,	/* we are using 'size = 4096,' so that native mode works with mtu >= 3800 */
```
Comment (giuseppelettieri) [inline sys/dev/netmap/netmap_mem2.c:527]: remove this comment

File: sys/dev/netmap/netmap_mem2.c [hunk 5]
```diff
@@ -571,7 +571,7 @@ struct netmap_mem_d nm_mem = {	/* Our memory allocator. */
 			.num  = 200,
 		},
 		[NETMAP_BUF_POOL] = {
-			.size = 2048,
+			.size = 2048,	/* again  we are using '.size = 4096,' for mtu >= 3800 */
```
Comment (giuseppelettieri) [inline sys/dev/netmap/netmap_mem2.c:574]: remove this comment

File: sys/dev/netmap/netmap_mem2.c [hunk 6]
```diff
@@ -1534,10 +1534,10 @@ netmap_mem_unmap(struct netmap_obj_pool *p, struct netmap_adapter *na)
 	struct netmap_lut *lut;
 	if (na == NULL || na->pdev == NULL)
 		return 0;
-	
+
```
Comment (giuseppelettieri) [inline sys/dev/netmap/netmap_mem2.c:1537]: please don't change the whitespace

**Review by giuseppelettieri — CHANGES_REQUESTED** [review]
Thanks. I have added a few inline comments. May you please fix and resubmit?

File: LINUX/configure (hunk 1 above)
Comment (feraralashkar) [reply LINUX/configure:940]: i think not possible since the Kconfig symbol for the driver is [ CONFIG_STMMAC_ETH ], while the driver directory is [ drivers/net/ethernet/stmicro/stmmac/ ]
and during configuration stage we get: following 
fashkar@barbarian:~/Downloads/EncDir/sperry/forked-netmap$ CONFIG_STMMAC=m CONFIG_E1000=m ./configure --driver-suffix=-netmap --no-ext-drivers --drivers=e1000,stmmac
...
********************************** WARNING **********************************
*** unknown driver: stmmac
*****************************************************************************

File: LINUX/netmap_linux.c (hunk 3 above)
Comment (feraralashkar) [reply LINUX/netmap_linux.c:1108]: need to check if  following function below can actually set the values correctly
static int stmmac_netmap_config(struct netmap_adapter *na,
				struct nm_config_info *info)

File: sys/dev/netmap/netmap_mem2.c (hunk 6 above)
Comment (feraralashkar) [reply sys/dev/netmap/netmap_mem2.c:1537]: Ack

File: sys/dev/netmap/netmap_mem2.c (hunk 5 above)
Comment (feraralashkar) [reply sys/dev/netmap/netmap_mem2.c:574]: Ack

File: sys/dev/netmap/netmap_mem2.c (hunk 4 above)
Comment (feraralashkar) [reply sys/dev/netmap/netmap_mem2.c:527]: Ack

File: LINUX/configure (hunk 1 above)
Comment (giuseppelettieri) [reply LINUX/configure:940]: OK, but you can add

```stmmac@conf := CONFIG_STMMAC_ETH```

to `LINUX/default-config.make.in_`. This will override the Kconfig symbol for just that driver.

Comment (feraralashkar) [reply LINUX/configure:940]: Ack

File: LINUX/netmap_linux.c (hunk 3 above)
Comment (feraralashkar) [reply LINUX/netmap_linux.c:1108]: ack

File: LINUX/configure (hunk 1 above)
Comment (feraralashkar) [reply LINUX/configure:940]: ack
